pkg/controller/batchrelease/labelpatch/patcher.go

```diff
@@ -55,59 +57,85 @@ func (r *realPatcher) PatchPodBatchLabel(ctx *batchcontext.BatchContext) error {
 
 // PatchPodBatchLabel will patch rollout-id && batch-id to pods
 func (r *realPatcher) patchPodBatchLabel(pods []*corev1.Pod, ctx *batchcontext.BatchContext) error {
-	// the number of active pods that has been patched successfully.
-	patchedUpdatedReplicas := int32(0)
-	// the number of target active pods that should be patched batch label.
-	plannedUpdatedReplicas := ctx.PlannedUpdatedReplicas
+	plannedUpdatedReplicasForBatches := r.calculatePlannedStepIncrements(ctx.Batches, int(ctx.Replicas), int(ctx.CurrentBatch))
+	var updatedButUnpatchedPods []*corev1.Pod
 
 	for _, pod := range pods {
+		if !pod.DeletionTimestamp.IsZero() {
+			klog.InfoS("Pod is being deleted, skip patching", "pod", klog.KObj(pod), "rollout", r.logKey)
+			continue
+		}
+		// we don't patch label for the active old revision pod
 		if !util.IsConsistentWithRevision(pod, ctx.UpdateRevision) {
+			klog.InfoS("Pod is not consistent with revision, skip patching", "pod", klog.KObj(pod), "rollout", r.logKey)
 			continue
 		}
-
-		podRolloutID := pod.Labels[v1beta1.RolloutIDLabel]
-		if pod.DeletionTimestamp.IsZero() && podRolloutID == ctx.RolloutID {
-			patchedUpdatedReplicas++
+		if pod.Labels[v1beta1.RolloutIDLabel] != ctx.RolloutID {
+			// for example: new/recreated pods
+			updatedButUnpatchedPods = append(updatedButUnpatchedPods, pod)
+			klog.InfoS("Find a pod to add updatedButUnpatchedPods", "pod", klog.KObj(pod), "rollout", r.logKey)
+			continue
 		}
-	}
 
-	// all pods that should be patched have been patched
-	if patchedUpdatedReplicas >= plannedUpdatedReplicas {
-		return nil // return fast
+		podBatchID, err := strconv.Atoi(pod.Labels[v1beta1.RolloutBatchIDLabel])
+		if err != nil {
+			klog.InfoS("Pod batchID is not a number, skip patching", "pod", klog.KObj(pod), "rollout", r.logKey)
+			continue
+		}
+		plannedUpdatedReplicasForBatches[podBatchID-1]--
 	}
-
-	for _, pod := range pods {
-		if pod.DeletionTimestamp.IsZero() {
-			// we don't patch label for the active old revision pod
-			if !util.IsConsistentWithRevision(pod, ctx.UpdateRevision) {
-				continue
+	klog.InfoS("updatedButUnpatchedPods amount calculated", "amount", len(updatedButUnpatchedPods), "rollout", r.logKey)
+	// patch the pods
+	for i := len(plannedUpdatedReplicasForBatches) - 1; i >= 0; i-- {
+		for ; plannedUpdatedReplicasForBatches[i] > 0; plannedUpdatedReplicasForBatches[i]-- {
+			if len(updatedButUnpatchedPods) == 0 {
+				return fmt.Errorf("no pods to patch for %v, batch %d", r.logKey, i+1)
 			}
-			// we don't continue to patch if the goal is met
-			if patchedUpdatedReplicas >= ctx.PlannedUpdatedReplicas {
-				continue
+			// patch the updated but unpatced pod
+			pod := updatedButUnpatchedPods[len(updatedButUnpatchedPods)-1]
+			clone := util.GetEmptyObjectWithKey(pod)
+			by := fmt.Sprintf(`{"metadata":{"labels":{"%s":"%s","%s":"%d"}}}`,
+				v1beta1.RolloutIDLabel, ctx.RolloutID, v1beta1.RolloutBatchIDLabel, i+1)
```

Comment:
can we make the batch id begin with 0, so as to make it consistent with the old behavior. changing the starting batch id may break some paas integrated with rollout.